src/runtime/pkg/containerd-shim-v2/service.go

```diff
@@ -794,7 +795,18 @@ func (s *service) Kill(ctx context.Context, r *taskAPI.KillRequest) (_ *ptypes.E
 		return empty, nil
 	}
 
-	return empty, s.sandbox.SignalProcess(spanCtx, c.id, processID, signum, r.All)
+	// kill(2) method can return ESRCH in certain cases, which is not handled by containerd cri server in container_stop.go.
+	// runc v2 translates ESRCH to ErrNotFound, so we'll handle similarly in Kata.
+	if err = s.sandbox.SignalProcess(spanCtx, c.id, processID, signum, r.All); err != nil &&
+		strings.Contains(err.Error(), "ESRCH: No such process") {
```

Comment:
Should this just be:
```
err == unix.ESRCH 
```

Reply:
@egernst I'm not sure if the err type is preserved across the grpc call between shim and kata agent. If that's the case, I'm happy to update it.

Reply:
Something similar to this has already been done on `WaitLocalProcess()`:
https://github.com/kata-containers/kata-containers/blob/358081c4ae1c9c8f03fb6520411f41ba3eda5f1f/src/runtime/virtcontainers/utils/utils.go#L323-L325

So, based on that, I think this approach is the way to go.
However, I wouldn't even warn here (or I would also warn there).  If this is a no-error, this is a no-error and users shouldn't be notified about this.

Apart from this, I do second Eric's opinion to at least check if we can use unix.ESRCH / syscall.ESRCH, we should be able to do that rather than matching strings.

Reply:
@fidencio `WaitLocalProcess` is to handle process exits on the host side, whereas this is trying to fix the race window that a container/process exits before a signal is processed by the agent.

Reply:
@fengwang666 While the race window is real, could you fold the change in `virtcontainers/container.go:signalProcess()` so that it works for all agent-handled signals?

Reply:
@bergwolf I like your suggestion. Updated.
@fidencio I personally think it's beneficial to log a warning on `ESRCH` error. It might indicate some other issues if it's happening more often than it should. So I added a log in the other place as well. Regarding checking the error with `unix.ESRCH / syscall.ESRCH`, I don't think ttrpc translates the rust ESRCH to golang ESRCH. Otherwise the error message we see in the log for [ESRCH](https://github.com/golang/go/blob/master/src/syscall/syscall_unix.go#L124) should be `errno 3`.

Reply:
> @fidencio WaitLocalProcess is to handle process exits on the host side, whereas this is trying to fix the race window that a container/process exits before a signal is processed by the agent.

Yet another reason to avoid pushing for reviews over the weekend. :-)

Reply:
> @fidencio I personally think it's beneficial to log a warning on ESRCH error. It might indicate some other issues if it's happening more often than it should.

The part that bothers me is we're working something around that should "just work" and then raising a red herring to folks debugging some problem that may or may not be related to this.
I sincerely don't think we should log a warn there, but you already have 2 approvals so I won't block it.

> Regarding checking the error with unix.[ESRCH](https://github.com/golang/go/blob/master/src/syscall/syscall_unix.go#L124) / syscall.ESRCH, I don't think ttrpc translates the rust ESRCH to golang ESRCH. Otherwise the error message we see in the log for ESRCH should be errno 3.

Fair. :-)